File: dannybrown/2021/src/day2.py
from helpers import read_string_puzzle_input

HORIZONTAL_DIRECTIONS = ['FORWARD']
VERTICAL_DIRECTIONS = ['UP', 'DOWN']
def day2_part1(instructions):
    horizontal_pos = 0
    vertical_pos = 0

    for instruction in instructions:
      direction, distance = instruction.split()

      if direction.upper() in HORIZONTAL_DIRECTIONS:
        horizontal_pos += int(distance)
      elif direction.upper() in VERTICAL_DIRECTIONS:
        if direction == 'up':
          vertical_pos -= int(distance)
        else:
          vertical_pos += int(distance)

    return horizontal_pos,  vertical_pos

def day2_part2(instructions):
    horizontal_pos = 0
    vertical_pos = 0
    aim = 0
    for instruction in instructions:
      direction, distance = instruction.split()

      if direction.upper() in HORIZONTAL_DIRECTIONS:
        horizontal_pos += int(distance)
        vertical_pos += aim * int(distance)
      elif direction.upper() in VERTICAL_DIRECTIONS:
        if direction == 'up':
          aim -= int(distance)
        else:
          aim += int(distance)

    return horizontal_pos, vertical_pos

if __name__ == '__main__':
  instructions = read_string_puzzle_input("dannybrown/input/day2.txt")

  horiz_1, vert_1 = day2_part1(instructions)
  horiz_2, vert_2 = day2_part2(instructions)

  print("Day 1 Part 1:", horiz_1 * vert_1)
  print("Day 2 Part 2:", horiz_2 * vert_2)

File: dannybrown/2021/src/day2/day2.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func getHorizontalDirections() []string {
	return []string{"FORWARD"}
}
func getVerticalDirections() []string {
	return []string{"UP", "DOWN"}
}

func check(e error) {
	if e != nil {
		panic(e)
	}
}

func contains(s []string, e string) bool {
	for _, a := range s {
			if a == e {
					return true
			}
	}
	return false
}

func day2_part1(instructions []string) (int, int) {
	var horizontalPos, verticalPos int

	for _, instruction := range instructions {
		instructionSlice := strings.Split(instruction, " ")

		direction := instructionSlice[0]
		distance, _ := strconv.Atoi(instructionSlice[1])

		if contains(getHorizontalDirections(), strings.ToUpper(direction)) {
			horizontalPos += distance
		} else if contains(getVerticalDirections(), strings.ToUpper(direction)) {
			if direction == "up" {
				verticalPos -= distance
			} else {
				verticalPos += distance
			}
		}
	}
	return horizontalPos, verticalPos
}

func day2_part2(instructions []string) (int, int) {
	var horizontalPos, verticalPos, aim int

	for _, instruction := range instructions {
		instructionSlice := strings.Split(instruction, " ")

		direction := instructionSlice[0]
		distance, _ := strconv.Atoi(instructionSlice[1])

		if contains(getHorizontalDirections(), strings.ToUpper(direction)) {
			horizontalPos += distance
			verticalPos += aim * distance
		} else if contains(getVerticalDirections(), strings.ToUpper(direction)) {
			if direction == "up" {
				aim -= distance
			} else {
				aim += distance
			}
		}
	}
	return horizontalPos, verticalPos
}

func main() {
	dat, err := os.Open("../input/day2.txt")
	check(err)
	defer dat.Close()
	scanner := bufio.NewScanner(dat)
	scanner.Split(bufio.ScanLines)
	var txtlines []string

	for scanner.Scan() {
		txtlines = append(txtlines, scanner.Text())
	}

	horiz1, vert1 := day2_part1(txtlines)
	horiz2, vert2 := day2_part2(txtlines)
	fmt.Println(horiz1*vert1) // Part 1
	fmt.Println(horiz2*vert2) // Part 2
}
